### ISSUE TITLE
[Ch 13][Notebook][Minor] 예상 소요 시간 "약 12분 (BERT 학습 ~10분)" — 실측 학습 49초

### ISSUE THREAD

**Issue by jwrim9305** [yoon-gu/neuqes-101#109]
## 대상

- **챕터**: Ch 13 — BERT Multi-label (Yelp 항목 키워드)
- **산출물**: Notebook (`13_bert_multilabel/13_bert_multilabel.ipynb` cell001), WikiDocs (`pages/13-bert_multilabel.md:5`), 책 (`book/chapters/ch13_bert_multilabel.tex`)
- **기준 커밋**: `eed3dbe`
- **환경**: Google Colab T4 / PyTorch 2.11.0+cu128, 전체 재실행 (2026-08-15)

## 실제 결과

챕터 도입부가 이렇게 적고 있습니다.

> **예상 소요 시간**: 약 12분 (BERT 학습 ~10분 + sklearn 비교 ~30초 + 평가/시각화)

실측은 **BERT 학습 49초** 입니다. `Trainer` 학습 표 헤더에 그대로 찍힙니다.

```
[626/626 00:49, Epoch 2/2]
```

`executed/13_bert_multilabel.ipynb` 의 같은 자리는 **00:37** 입니다.

| | 본문 주장 | `executed/` | 재실행 |
|---|---|---|---|
| BERT 학습 | ~10분 | 00:37 | **00:49** |

**12배 이상 과대**입니다. 626 스텝 / batch 16 / `max_length=128` / fp16 인 T4 워크로드라 10분이 나올 수 있는 규모가 아닙니다.

## 기대 결과

**문장 구조는 그대로 두고 숫자만 실측에 맞춥니다.**

```diff
- **예상 소요 시간**: 약 12분 (BERT 학습 ~10분 + sklearn 비교 ~30초 + 평가/시각화)
+ **예상 소요 시간**: 약 5분 (BERT 학습 약 1분 + sklearn 비교 약 30초 + 평가/시각화)
```

`약 5분` 은 모델 다운로드(268MB)와 데이터셋 로드를 포함한 값입니다. 학습 자체는 49초였습니다.

**세부 항목을 빼고 상한만 적는 방식도 좋습니다.** 소요 시간은 하드웨어·네트워크에 좌우되므로 분 단위로 쪼개 적을수록 틀리기 쉽습니다.

```markdown
**예상 소요 시간**: 약 5분 이내 (T4 GPU 기준)
```

어느 쪽이든 **전 챕터가 같은 방식을 쓰는 것** 이 중요합니다 — 아래 참고 항목을 봐 주세요.

> 물결표를 `약` 으로 바꾼 것은 같은 줄을 다루는 별건(FAQ Q5 이슈)의 취소선 렌더링 정정과 표기를 맞추기 위해서입니다. 이 줄 자체는 취소선이 생기지 않습니다.

## 재현 절차

1. Colab T4 에서 노트북을 전체 실행합니다.
2. cell 21 학습 셀의 진행 표시줄 헤더 `[626/626 00:49, Epoch 2/2]` 를 확인합니다.
3. `executed/13_bert_multilabel.ipynb` 의 같은 셀 HTML 출력에서 `[626/626 00:37, ...]` 을 확인합니다.

## 전파 범위

산문만 고치면 되고 **재실행은 불필요**합니다.

- `13_bert_multilabel/13_bert_multilabel.ipynb` cell001 (도입부)
- `pages/13-bert_multilabel.md:5`
- `book/chapters/ch13_bert_multilabel.tex` 도입부 → EPUB 재빌드

## 다른 챕터에도 같은 패턴이 있습니다 — 함께 처리하기를 제안합니다

Ch 13 만의 문제가 아닙니다. **BERT 파인튜닝 계열 챕터가 모두 같은 문구 틀** 을 쓰고 있고, 실측(`executed/` 의 `Trainer` 학습 표 헤더)과 대조하면 전부 어긋납니다.

| 챕터 | 본문 주장 | 실측 학습 | 담당 |
|---|---|---|---|
| 11 | 방식 B `~8분` + 방식 A `~8분` (총 20분) | **00:29 + 00:30** | 페어 A |
| 12 | 학습 `~10분` (총 12분) | 00:37 | 페어 B ([#102](https://github.com/yoon-gu/neuqes-101/issues/102)) |
| **13** | 학습 `~10분` (총 12분) | **00:37** (재실행 00:49) | **본 이슈** |
| 14 | ON `~9분` + OFF `~9분` (총 22분) | **00:38 + 00:40** | 페어 B |
| 15 | 2 에폭 학습 `~10분` (총 12분) | **00:46** | 페어 B |
| 16 | 2 에폭 학습 `~10분` (총 13분) | **00:39** | 페어 B |
| 17 | 2 에폭 학습 `-10분` (총 13분) | **00:40** | 페어 B |
| 18 | ON `약 10분` + baseline `약 10분` (총 22분) | **00:40 + 00:40** | 페어 B |
| 21 | MLM 3 epoch `약 8-10분` | **00:14 + 00:14** | 페어 C |
| 23 | MLM 3 epoch `약 8-10분` | **00:10 + 00:12** | 페어 C |

GPT 계열은 대체로 맞습니다 — Ch 25(`25-30분` / 19:02), Ch 27(`25-30분` / 17:05), Ch 28(`10-15분` / 02:22). **BERT 파인튜닝 챕터에만 몰려 있습니다.**

확인 방법:

```bash
python3 - <<'PY'
import json, re, pathlib
def claim(nb):
    for c in json.load(open(nb))['cells'][:4]:
        if c['cell_type'] != 'markdown': continue
        for l in ''.join(c['source']).split('\n'):
            if '소요 시간' in l: return l.strip()
def measured(p):
    if not p.exists(): return []
    return [m.group(2) for c in json.load(open(p))['cells'] for o in c.get('outputs', [])
            for m in re.finditer(r'\[\s*(\d+)/\1\s+(\d+:\d+(?::\d+)?)',
                                 ''.join(o.get('data', {}).get('text/html', [])))]
for src in sorted(pathlib.Path('.').glob('[0-9][0-9]_*/[0-9][0-9]_*.ipynb')):
    c = claim(src)
    if c: print(f"{src.name:<32} {c.split(':',1)[1].strip()[:48]:<50} {measured(pathlib.Path('executed')/src.name)}")
PY
```

**한 번에 정리하는 편이 낫습니다.** 챕터마다 따로 고치면 표기 방식이 또 갈립니다. 위에 제안한 두 형태(세부 항목 유지 / `약 N분 이내` 로 단순화) 중 하나를 정해 전 챕터에 적용하기를 제안합니다.

> **Ch 17·19 는 물결표가 아니라 하이픈입니다** — `캐시 -10s + 2 에폭 학습 -10분`, `데이터 다운로드 -2분`. "빼기" 로 읽히므로 함께 고치면 좋겠습니다.

## 심각도

**Minor 로 제안합니다** — Ch 12 의 동일 건 [#102](https://github.com/yoon-gu/neuqes-101/issues/102)(`예상 소요 시간 "BERT 학습 ~10분" — 실측은 37초`)와 등급을 맞췄습니다. 소요 시간은 **하드웨어·네트워크·캐시 상태에 좌우되는 값** 이고, 읽는 데 지장이 없으며 학습 결과에도 영향이 없습니다.

> 참고 — `fix/ch12-review-2026-09` 브랜치의 `.claude/skills/chapter-review/SKILL.md` 는 이 유형을 *"'약 12분' 인데 실측 37초 → **결함 아님** (런타임·네트워크에 좌우됨)"* 으로 분류하고 있습니다. 실행 환경에 좌우되는 값을 노이즈로 걸러내자는 취지에 동의하며, 등급을 Minor 로 둔 것도 같은 이유입니다. 다만 최소 10개 챕터가 일관되게 어긋나 있어 **표기 방식 자체를 한 번 정리할 가치는 있다** 고 보아 등록합니다. 결함으로 다룰지, 표기 규약 논의([#63](https://github.com/yoon-gu/neuqes-101/issues/63))로 넘길지는 재검수자와 맞추면 좋겠습니다.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자 (@jwrim9305)
- 연결: 챕터 검수 이슈 [#37](https://github.com/yoon-gu/neuqes-101/issues/37) / 동일 유형 [#102](https://github.com/yoon-gu/neuqes-101/issues/102)
- **전 챕터를 한 번에 정리하는 편을 권합니다** — 위 참고표대로 최소 10개 챕터가 같은 문구 틀을 쓰고 있습니다. 표기 방식이 정해지면 본 이슈와 #102 는 그 결정에 따라 함께 닫으면 됩니다.

**Comment by jwrim9305** [comment]
## 수정했습니다

브랜치 `fix/ch13-review-2026-09-b1` · 커밋 `c001c3b`

```diff
- **예상 소요 시간**: 약 12분 (BERT 학습 ~10분 + sklearn 비교 ~30초 + 평가/시각화)
+ **예상 소요 시간**: 약 5분 (BERT 학습 약 1분 + sklearn 비교 약 30초 + 평가/시각화)
```

**세 곳을 함께 고쳤습니다.**

| 파일 | |
|---|---|
| `13_bert_multilabel/13_bert_multilabel.ipynb` | 도입부 |
| `pages/13-bert_multilabel.md:5` | |
| `13_bert_multilabel/README.md:24` | `약 12분 (BERT ~10분 …)` → `약 5분 (BERT 학습 약 1분 …)` |

README 는 이슈 본문의 전파 범위에 없었는데, 같은 문구가 있어 함께 정정했습니다.

물결표는 `약` 으로 바꿨습니다 — [#60](https://github.com/yoon-gu/neuqes-101/issues/60) 이후 Ch 20 이 쓰는 근사 표기 관행에 맞추고, 괄호와 붙었을 때 취소선으로 렌더링되는 것([#107](https://github.com/yoon-gu/neuqes-101/issues/107))도 피합니다.

총 시간 `약 5분` 은 모델 다운로드(268MB)와 데이터셋 로드를 포함한 값이라 환경에 따라 달라집니다. 이 이슈에서 확정한 것은 **학습 구간이 분 단위가 아니라 초 단위** 라는 점입니다.

### 재실행·재생성 필요 여부

| 산출물 | 필요 | 사유 |
|---|---|---|
| **노트북 실행** (`executed/`) | ❌ 불필요 | 도입부는 마크다운 셀 |
| **WikiDocs** (`pages/`) | ✅ 이미 반영 | `13-bert_multilabel.md` 동시 수정 |
| **EPUB** (`tex`) | ✅ 필요 | `notebook_to_tex.py --chapters 13 --use-executed` → 빌드 |

> 참고 — 같은 문구 틀을 쓰는 챕터가 최소 10개입니다(Ch 11·12·13·14·15·16·17·18·21·23). 표기 방식을 정해 한 번에 정리하는 편이 낫다고 보아 이슈 본문에 대조표를 넣어 두었습니다.

**Comment by sy2399** [comment]
## 재검토 — 2차 검수자 @sy2399

**주장 유효**이고 **Minor 유지**에 동의합니다.

- 제시하신 스크립트를 돌려 표를 재현했습니다 — Ch 11 `00:29+00:30`, Ch 14 `00:38+00:40`, Ch 15 `00:46`, Ch 16 `00:39`, Ch 17 `00:40`, Ch 18 `00:40+00:40`, Ch 21 `00:14+00:14`, Ch 23 `00:10+00:12`. Ch 17·19 의 하이픈(`-10분`) 표기도 실재합니다.
- 소요 시간을 결함으로 다룰지에 대해서는 **검수 대상이 맞습니다** (2026-08-21 확정, 이전 "결함 아님" 판정은 뒤집혔습니다). 학습자가 도입부를 보고 세션을 계획하는데 배수 단위로 어긋나기 때문입니다.

**수정 확인** — `fix/ch13-review-2026-09-b1` `c001c3b`. `약 5분 (BERT 학습 약 1분 + sklearn 비교 약 30초 + 평가/시각화)` 로 정정됐고, 노트북 · `pages/13-bert_multilabel.md` · **챕터 `README.md:24`** 세 곳 모두 반영됐습니다. 이슈 전파 목록에 없던 챕터 README 까지 잡혀 있습니다.

## 세 축 확인 후 닫습니다

| 축 | 상태 |
|---|---|
| 📓 Notebook | 수정됨 (`c001c3b`, 챕터 README 포함) |
| 📄 WikiDocs | `pages/13-bert_multilabel.md` 수정됨 |
| 📕 EPUB | **해당 없음** — compact 원고에 도입부 소요 시간 문구가 없습니다 |

> 표기 한 가지만 남깁니다. Ch 12 는 `약 3분 (BERT 학습 약 40초 …)`, Ch 13 은 `약 5분 (BERT 학습 약 1분 …)` 인데 두 챕터 모두 626 step / `00:37` 로 워크로드가 같습니다. 전 챕터 표기 통일은 #63 에서 함께 맞추면 좋겠습니다.

**Comment by jwrim9305** [comment]
## 파이프라인 완료 — 다만 EPUB 에는 실리지 않습니다

브랜치 `fix/ch13-review-2026-09-merged` · 재생성 `2a21d3a`

「예상 소요 시간」이 있던 **도입부 셀(cell000) 전체가 압축 원고에서 대체**됩니다. `--compact-code` 로 재생성하면 이 한 줄만 남습니다.

```latex
\chaptermeta{13_bert_multilabel/13_bert_multilabel.ipynb}{<Colab URL>}{Yelp 항목 합성 라벨을 BERT의 라벨별 sigmoid와 BCE로 학습}
```

`book/designsystem.sty:475`·`:303` 에서 **Chapter map 박스 + QR 코드 박스**로 펼쳐집니다. 「목표」·「환경」·「예상 소요 시간」·「학습 흐름」이 모두 빠지고 *"이 장의 초점"* 한 줄로 압축됩니다.

| 산출물 | 상태 |
|---|---|
| **노트북 실행** (`executed/`) | ❌ 불필요 (마크다운 셀) — #105 재실행 때 함께 갱신 |
| **WikiDocs** (`pages/`) | ✅ 완료 — `2a21d3a` |
| **EPUB** (`tex`) | **해당 없음** — 도입부가 `\chaptermeta` + QR 로 대체 |

앞 댓글의 「EPUB 필요」 표기를 정정합니다.

> 참고 — 이 이슈에서 제기한 **전 챕터 소요 시간 표기 정리**는 별개로 남아 있습니다. 압축 원고가 도입부를 걷어내므로 **책에는 영향이 없고, 노트북·WikiDocs 축에서만 유효한 논의**가 됩니다.
